packages/e2e/lib/e2e.dart

```diff
@@ -184,4 +241,88 @@ class E2EWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding {
     );
     results[description] ??= 'success';
   }
+
+  /// Handle commands from WebDriver tests.
+  ///
+  /// WebDriver provides a set methods (ex: screenshot, resize, refresh) which
+  /// are useful for writing end-to-end tests on multiple browsers.
+  ///
+  /// e2e tests are sending [TestStatus] messages to state that they are ready
+  /// to run a WebDriver action ([TestStatus.wait_on_webdriver_command]) or
+  /// they are finished a web driver action
+  /// ([TestStatus.webdriver_command_complete])
+  ///
+  /// The tests running on the browser sends request these WebDriver operations
+  /// from e2e tests via replying with [WebDriverAction]s.
+  Future<Map<String, dynamic>> _callbackForWebDriver(
+      Map<String, String> params) async {
+    final String command = params['command'];
+    // Test status is added as an exta message.
+    final String extraMessage = params['message'] ?? '';
+    Map<String, String> response;
+    switch (command) {
+      case 'request_data':
+        // If not extra message is added continue as normal.
+        if (extraMessage.isEmpty) {
+          final bool allTestsPassed = await _allTestsPassed.future;
+          response = <String, String>{
+            'message': allTestsPassed
+                ? Response.allTestsPassed(data: reportData).toJson()
+                : Response.someTestsFailed(
+                    _failureMethodsDetails,
+                    data: reportData,
+                  ).toJson(),
+          };
+        } else {
+          // If Test status is `wait_on_webdriver_command` send the first
+          // command in the `commandPipe` to the tests.
+          if (extraMessage == '${TestStatus.wait_on_webdriver_command}') {
```

Comment:
Why would the driving script do this? Shouldn't the test itself be awaiting whatever it needs (i.e. taking the screenshot), and the driving script just waits for the test reporting to finish?

Another way of putting this concern: how does the caller know which web driver command(s) it's waiting on, that there won't be another one it's actually interested in coming right after the one it just waited on?

Reply:
_Why would the driving script do this? Shouldn't the test itself be awaiting whatever it needs (i.e. taking the screenshot), and the driving script just waits for the test reporting to finish?_

Only driver script can take screenshots. Almost all browsers implements the webdriver API which allows us to take screenshot. We can also extend this to do other import things such as changing screensize, contorrling cookies: https://www.w3.org/TR/webdriver1/.  Since only driver test can perform these operations, we developed this handshake mechanism between the integration_test and the driver test. 

In summary: We are sending the request from the integration_test (ex: `WebDriverActionTypes.screenshot`), waiting on the result from the driver. If there are multiple test methods which are asking for a `WebDriverAction`, they wait on the mutex, driver handles them one by one.

_Another way of putting this concern: how does the caller know which web driver command(s) it's waiting on, that there won't be another one it's actually interested in coming right after the one it just waited on?_

That is why I used a synchronization method (such as mutex) to send these requests to driver **one at a time**. I tested with many scenarios:
1) requests in within different methods.
2) few request in a few methods.
3) Failing vs passing tests in different methods.

This handshake approach worked well.